File: src/linked_lists/java/Intersection.java
package linked_lists.java;

import static linked_lists.java.Palindrome.arrayToLinkedList;
import static linked_lists.java.Palindrome.getLength;

import java.util.HashSet;
import java.util.Set;

/**
 *
 */
public class Intersection {

  public static void main(String[] args) {
    Node list1  = arrayToLinkedList(new int[]{1, 2, 3});
    Node list2  = arrayToLinkedList(new int[]{11, 22});
    Node common = arrayToLinkedList(new int[]{4, 5, 6});
    list1 = concat(list1, common);
    list2 = concat(list2, common);

    System.out.println("Method 1 (find circle):");
    Node node = intersection(list1, list2);
    System.out.println(
        node == null ? "No intersection" : "Data of intersection node is " + node.data);

    System.out.println("Method 2 (hash set):");
    node = intersectionHashSet(list1, list2);
    System.out.println(
        node == null ? "No intersection" : "Data of intersection node is " + node.data);

    System.out.println("Method 3 (alignment):");
    node = intersectionAlignment(list1, list2);
    System.out.println(
        node == null ? "No intersection" : "Data of intersection node is " + node.data);

    // There are also other methods, such as using two stacks.
  }

  private static Node concat(Node head1, Node head2) {
    if (head1 == null) {
      return head2;
    }
    Node ptr = head1;
    while (ptr.next != null) {
      ptr = ptr.next;
    }
    ptr.next = head2;
    return head1;
  }

  private static Node intersection(Node head1, Node head2) {
    if (head1 == null || head2 == null) {
      return null;
    }

    Node ptr = head1;
    while (ptr.next != null) {
      ptr = ptr.next;
    }
    final Node tail1 = ptr;

    /* only one node in list1 */
    if (tail1 == head1) {
      return head1 == head2 ? head1 : null;
    }

    tail1.next = head2;

    /* find the entrance of circle in list1 */
    ptr = head1;
    Node fast = head1;
    while (fast != null && fast.next != null) {
      ptr = ptr.next;
      fast = fast.next.next;

      if (ptr == fast) {
        break; // list1 has a circle
      }
    }
    ptr = head1;
    while (ptr != fast) {
      ptr = ptr.next;
      assert fast != null;
      fast = fast.next;
    }
    tail1.next = null;

    return ptr;
  }

  private static Node intersectionHashSet(Node head1, Node head2) {
    Set<Node> set = new HashSet<>();
    for (Node ptr = head1; ptr != null; ptr = ptr.next) {
      set.add(ptr);
    }
    for (Node ptr = head2; ptr != null; ptr = ptr.next) {
      if (set.contains(ptr)) {
        return ptr;
      }
    }
    return null;
  }

  private static Node intersectionAlignment(Node head1, Node head2) {
    int  len1      = getLength(head1);
    int  len2      = getLength(head2);
    Node shortList = len1 < len2 ? head1 : head2;
    Node longList  = len1 < len2 ? head2 : head1;

    int len = Math.max(len1, len2);
    int dis = Math.abs(len2 - len1);
    for (int i = 0; i < len; i++) {
      longList = longList.next;
      if (i < dis) {
        continue;
      }

      shortList = shortList.next;
      if (shortList == longList) {
        return shortList;
      }
    }

    return null;
  }
}

File: src/linked_lists/kotlin/Intersection.kt
package linked_lists.kotlin

import kotlin.math.abs
import kotlin.math.max

/**
 *
 */
fun main(args: Array<String>) {
    var list1 = arrayToLinkedList(intArrayOf(1, 2, 3))
    var list2 = arrayToLinkedList(intArrayOf(11, 22))
    val common = arrayToLinkedList(intArrayOf(4, 5, 6))
    if (list1 != null && list2 != null && common != null) {
        list1 = concat(list1, common)
        list2 = concat(list2, common)

        println("Method 1 (find circle):")
        var node = intersection(list1, list2)
        println(if (node == null) "No intersection" else "Data of intersection node is " + node.data)

        println("Method 2 (hash set):")
        node = intersectionHashSet(list1, list2)
        println(if (node == null) "No intersection" else "Data of intersection node is " + node.data)

        println("Method 3 (alignment):")
        node = intersectionAlignment(list1, list2)
        println(if (node == null) "No intersection" else "Data of intersection node is " + node.data)
    }
}

fun concat(head1: Node, head2: Node): Node {
    var ptr = head1
    while (ptr.next != null) {
        ptr = ptr.next!!
    }
    ptr.next = head2
    return head1
}

fun intersection(list1: Node, list2: Node): Node? {
    var ptr = list1 as? Node
    while (ptr != null && ptr.next != null) {
        ptr = ptr.next
    }
    val tail = ptr

    tail!!.next = list2
    var fast = list1 as? Node
    ptr = list1
    while (fast != null && fast.next != null) {
        fast = fast.next!!.next
        ptr = ptr!!.next

        if (ptr == fast) break
    }
    ptr = list1
    while (ptr != null && fast != null) {
        if (ptr == fast) break
        ptr = ptr.next
        fast = fast.next
    }

    tail.next = null
    return ptr
}

fun intersectionHashSet(list1: Node, list2: Node): Node? {
    val set = HashSet<Node>()
    var ptr = list1 as? Node
    while (ptr != null) {
        set.add(ptr)
        ptr = ptr.next
    }
    ptr = list2
    while (ptr != null) {
        if (set.contains(ptr)) return ptr
        ptr = ptr.next
    }

    return null
}

fun intersectionAlignment(list1: Node, list2: Node): Node? {
    val len1 = getLength(list1)
    val len2 = getLength(list2)

    var shortList: Node? = if (len1 < len2) list1 else list2
    var longList: Node? = if (len1 < len2) list2 else list1
    val len = max(len1, len2)
    val dis = abs(len2 - len1)
    for (i in 1..len) {
        longList = longList!!.next
        if (i in 1..dis) continue
        shortList = shortList!!.next

        if (longList == shortList) return shortList
    }

    return null
}
